multiapps-controller-process/src/main/java/org/cloudfoundry/multiapps/controller/process/steps/DetermineServiceCreateUpdateServiceActionsStep.java

```diff
@@ -224,8 +224,12 @@ private CloudServiceInstanceExtended mergeCredentials(CloudServiceInstanceExtend
                                                     .withCredentials(result);
     }
 
-    private boolean shouldUpdateKeys(CloudServiceInstanceExtended service, List<CloudServiceKey> serviceKeys) {
-        return !(service.isUserProvided() || CollectionUtils.isEmpty(serviceKeys));
+    private boolean shouldUpdateKeys(CloudServiceInstanceExtended service, List<CloudServiceKey> serviceKeys, ProcessContext context) {
+        return !(service.isUserProvided() || CollectionUtils.isEmpty(serviceKeys) && shouldNotDeleteServiceKeys(context));
```

Comment:
Due to the moderate complexity of this statement, it's a bit hard to discern the result.
It would be clearer if the 3rd expression is outside the initial parenthesis.
As far as I understand, right now, this would return true (as in, the keys should be updated) when the service is not user-provided or has at least one service key and its service keys _should_ be deleted (because the new function returns true if service keys should not be deleted and we are inverting that result).
Is this the correct behaviour?

Reply:
I agree that the statement is rather complicated and I will try to simplify it with your suggestion.  
As for the behaviour, the method would currently return true (meaning the keys should be updated) when the service is not user provided and either there is at least one service key or service keys should be deleted.